packages/material-ui/src/Stepper/Stepper.js

```diff
@@ -52,24 +53,24 @@ const Stepper = React.forwardRef(function Stepper(props, ref) {
       completed: false,
       disabled: false,
     };
-
+    
     if (activeStep === index) {
       state.active = true;
     } else if (!nonLinear && activeStep > index) {
       state.completed = true;
     } else if (!nonLinear && activeStep < index) {
       state.disabled = true;
     }
-
+    
     return React.cloneElement(step, {
```

Comment:
Using context should allow us to remove cloneElement usage too. Some extra work might be needed here.

Reply:
Thanks, I thought about it, the only reason I didn't remove it is the `index` prop. How do we pass the `index` to the `Step` without  `Children.map` + `cloneElement` ?

Reply:
Hmm, I'd have to take a close look. We might be able to leverage [descendants](https://github.com/mui-org/material-ui/blob/154c8f1fbc3e8199d8dfbbda0e48688f9c3501a4/packages/material-ui-lab/src/TreeView/descendants.js) here. Might be able to do something simpler.

Reply:
@joshwooding Would such an approach work with the initial server-side render?

Reply:
@oliviertassinari I took out the code that allowed it but you had to pass an explicit index: https://github.com/reach/reach-ui/blob/86a046f54d53b6420e392b3fa56dd991d9d4e458/packages/descendants/src/index.tsx#L43

Reply:
Alright so we face a hard tradeoff here. With two options:

- We keep the cloneElement to managing the index. If we do such, what's the use case for using the context, can't we just say props have to be spread, end of the story.
- We ask users to provide an explicit index. In which case, we don't need the descendant logic.

Is this accurate?

Reply:
Yeah, although last is only used to apply a class in StepContent.

Reply:
And we make the index a public API? So that worth case, people can manually add it if they break the default cloneElement? Sounds like a great tradeoff.

Reply:
Warning: Slightly off-topic - I apologise but still is very relevant :P 

Yeah, I might even re-visit TreeView and do the same. Problems I current see with useDescendant:

Nodes have to be in the DOM which causes issues for virtualization (mainly selection) and if we wanted to separate the component API from the DOM rendering component in the future.

compareDocumentPosition is quite slow:

![image](https://user-images.githubusercontent.com/12938082/85961940-cb504800-b9a5-11ea-8085-98b99c175520.png)

This was a test after running `docs:build` and opening a node with 1,000 children it shows the top things taking up time sorted by self-time. You can think of `y` representing the set of descendants.  Part of this would be avoided with separating the DOM representation and the component representation but the new descendants API makes this difficult due to needing an element as stated above.

As for comparison providing `index` via a prop:

![image](https://user-images.githubusercontent.com/12938082/85962268-cf7d6500-b9a7-11ea-9483-33af2734ad93.png)

Reply:
Interesting. I had virtualization in the back of my mind for the enterprise version. It sounds like quite a challenge.

Reply:
It will be 🤯 there are ways to make it simpler - treat the tree like one massive list and lose the Dom structure and encounter styling issues but it’s something that I try to keep in mind when working on the tree.

Reply:
Every time I'm playing with the VSCode file tree, I'm amazed by it.

Reply:
@oliviertassinari @joshwooding  To clarify - we are keeping `cloneElement` only for the `index` value and also adding `index` prop to the `Step` component?

Reply:
Also, the `last`  prop depends on `index` and children lengths, we could either pass it along with the index through `cloneElement` or pass children length through the `Stepper` context.

Reply:
Thanks for raising last, yeah, we could make it public too.

Reply:
@oliviertassinari and we are keeping `cloneElement` for the `index` and `last` ?

Reply:
From what I understood, yes. There is no way around it